File: leetcode/Q121_BestTimetoBuyandSellStock.py
from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_num = 0
        cache = [0]
        for i in range(1, len(prices)):
            cache.append(max(0, cache[i-1]-(prices[i-1]-prices[i])))
            max_num = max(max_num, cache[i])
        return max_num

File: src/easy/Q121_BestTimetoBuyandSellStock.java
package easy;

public class Q121_BestTimetoBuyandSellStock {
    class Solution {
        public int maxProfit(int[] prices) {
            int min = prices[0];
            int max = 0;
            for(int p : prices){
                min = Math.min(min, p);
                max = Math.max(max, p - min);
            }
            return max;
        }
    }
}
